src/macaron/slsa_analyzer/ci_service/github_actions/analyzer.py

```diff
@@ -0,0 +1,718 @@
+        source_path: str,
+        parsed_obj: dict,
+        caller: BaseNode | None = None,
+        model: ThirdPartyAction | None = None,
+    ) -> None:
+        """Initialize instance.
+
+        Parameters
+        ----------
+        name : str
+            Name of the workflow (or URL for reusable and external workflows).
+        node_type : GHWorkflowType
+            The type of workflow.
+        source_path : str
+            The path of the workflow.
+        parsed_obj : dict
+            The parsed Actions workflow object.
+        caller: BaseNode | None
+            The caller node.
+        model: ThirdPartyAction | None
+            The static analysis abstraction for the third-party GitHub Action.
+        """
+        super().__init__(caller=caller)
+        self.name = name
+        self.node_type: GHWorkflowType = node_type
+        self.source_path = source_path
+        self.parsed_obj = parsed_obj
+        self.model = model
+
+    def __str__(self) -> str:
+        return f"GitHubWorkflowNode({self.name},{self.node_type})"
+
+
+class GitHubJobNode(BaseNode):
+    """This class is used to create a call graph node for GitHub Actions jobs."""
+
+    def __init__(self, name: str, source_path: str, parsed_obj: dict, caller: BaseNode) -> None:
+        """Initialize instance.
+
+        Parameters
+        ----------
+        name : str
+            Name of the workflow (or URL for reusable and external workflows).
+        source_path : str
+            The path of the workflow.
+        parsed_obj : dict
+            The parsed Actions workflow object.
+        caller: BaseNode
+            The caller node.
+        """
+        super().__init__(caller=caller)
+        self.name = name
+        self.source_path = source_path
+        self.parsed_obj = parsed_obj
+
+    def __str__(self) -> str:
+        return f"GitHubJobNode({self.name})"
+
+
+def find_expression_variables(value: str, exp_var: str) -> Iterable[str]:
+    """Find all the matching GitHub Actions expression variables in a string value.
+
+    GitHub Actions Expression syntax: ${{ <expression> }}
+    See https://docs.github.com/en/actions/learn-github-actions/expressions#about-expressions
+
+    Parameters
+    ----------
+    value: str
+        The value in which the expression values are searched.
+    exp_var: str
+        The expression variable name.
+
+    Yields
+    ------
+    Iterable[str]
+        The expression variable names.
+
+    Examples
+    --------
+    >>> list(find_expression_variables("echo ${{ inputs.foo }}", "inputs"))
+    ['foo']
+    >>> list(find_expression_variables("echo ${{ inputs.foo }} ${{ inputs.bar }}", "inputs"))
+    ['foo', 'bar']
+    >>> list(find_expression_variables("echo ${{ inputs.foo }} ${{ inputs.bar }}", "matric"))
+    []
+    """
+    expressions = re.findall(r"\$\{\{.*?\}\}", value)
+    for exp in expressions:
+        match = re.match(r"\$\{\{\s+" + exp_var + r"\.(?P<variable>(.*?))\s+\}\}", exp)
```

Comment:
Does the regex pattern `r"\$\{\{\s+" + exp_var + r"\.(?P<variable>(.*?))\s+\}\}"` need to be computed for every loop?

Reply:
I'm not if I understand this question, but this loop is needed if that's what you mean.

Reply:
Oh sorry if my wordings was a bit confusing. 
I meant to say that the string `r"\$\{\{\s+" + exp_var + r"\.(?P<variable>(.*?))\s+\}\}"` seems to not depend on the loop variable `exp` so I think we can compute it outside of the for loop and reuse it everytime
```
    exp_var_str = r"\$\{\{\s+" + exp_var + r"\.(?P<variable>(.*?))\s+\}\}"
    for exp in expressions:
        match = re.match(exp_var_str, exp)
```

Reply:
If you can hoist the regex string out of the loop you might as well re.compile the regex out of the loop as well, since that may have some performance benefit.

Reply:
Yes, but that would make the regex complicated.

Reply:
See commit [6832dcd](https://github.com/oracle/macaron/pull/673/commits/6832dcd2d8a32a57558360b6c090ebdd8e7039a8)